Listing APIs

JonathonYY opened the pull request:
Initial implementations; beginning to do unit tests via jest

JonathonYY committed:
Implement rate API

JonathonYY committed:
update rate to better use firebase

JonathonYY committed:
initial report implementation

JonathonYY committed:
Implement listing delete API

JonathonYY committed:
Update timestamp on listing change

JonathonYY committed:
Update interfaces

JonathonYY committed:
Begin defining rate unit tests

Edwinhe03 commented on src/lib/firebase/firestore/types.ts:
```diff
@@ -44,7 +47,9 @@ export function newListing(): Listing {
-    "id": "",
+    "owner_pfp": "",
+    "owner_name": "",
+    seller_rating: 0
```
missing quotes: `"seller_rating"`

JonathonYY replied:
Fixed locally, will be included in next push

JonathonYY committed:
Complete rate API unit tests

JonathonYY committed:
Split deleteListing logic to separate function

JonathonYY committed:
Switch to log at warning level

JonathonYY committed:
fix minor typo

JonathonYY committed:
Unit tests for delete listing

JonathonYY committed:
Update report API

Only allow valid users to report a listing
Enforce timeout between reports
Automatically delete listing once it has been reported a
certain number of times

JonathonYY committed:
Unit tests for report listing

Edwinhe03 commented on src/lib/firebase/firestore/listing/deleteListing.test.ts:
```diff
@@ -0,0 +1,166 @@
+  return {
+    ...jest.requireActual('firebase/firestore'),
+    doc: jest.fn((db, table, id) => {
+      if (!db[table[id]]) { // if doc doesn't exist, return ref anyway
```
nit: `!db[table][id]` for consistency

Edwinhe03 commented on src/app/api/listing/[listing_id]/rate/route.ts:
```diff
@@ -18,13 +21,92 @@ export async function PATCH(
+
+      const buyerData = buyerSnapshot.data() as User;
+      if (buyerData) {
+        if (user_id in ratings) {
```
For both buyer and seller ratings, in what context are they able to go back and change their rating? Also it seems that they only "complete" their purchase/sale respectively when the other party rates them. Is this what we want? I'm not exactly sure what frontend wants the user journey to look like here @eliu0120 @RyanWang3

JonathonYY replied:
My understanding was that buyer/seller ratings could be adjusted until the listing was deleted, but that can be changed

JonathonYY replied:
Another option: Users *cannot* change their ratings, and the `rate` API automatically deletes the listing once both buyer and seller have provided a rating

Edwinhe03 replied:
gotcha. anyways, having the api be more powerful than needed is not a bad thing, so let's keep it

Edwinhe03 commented on src/app/api/listing/[listing_id]/rate/route.ts:
```diff
@@ -18,13 +21,92 @@ export async function PATCH(
 
-  return NextResponse.json({ data: { listing_id: listing_id }, error: null });
+    // Reference to listing in firestore
+    const listingRef = doc(db, 'listings', listing_id);
```
Not needed for this PR, but once @zhou-michael and my CRUD API/util functions are merged, we can replace these GET listing/user lines with our util functions (ie. lines 37-47).

Edwinhe03 commented on src/app/api/listing/[listing_id]/report/route.ts:
```diff
@@ -15,13 +22,63 @@ export async function PATCH(
+    const listing_id = (await params).listing_id;
+
+    // get reporter data from req body
+    const {user_id} = await req.json();
```
might be worthwhile to add basic input validation to check that `user_id` is not undefined rather than implicitly letting `getDoc` throw an error for not finding the user later

Edwinhe03 commented on src/app/api/listing/[listing_id]/route.ts:
```diff
@@ -72,10 +73,21 @@ export async function DELETE(
+    const listing_id = (await params).listing_id;
+
+    // get user id from req body
+    const {user_id} = await req.json();
```
ditto comment on validation from report API

Edwinhe03 commented on src/app/api/listing/[listing_id]/report/route.ts:
```diff
@@ -15,13 +22,63 @@ export async function PATCH(
+    if (reporters.length + 1 >= removal_threshold) {
+      await deleteListing(listing_id, owner);
+    } else {
+      await updateDoc(listingRef, { reporters: arrayUnion(user_id), updated: serverTimestamp() });
```
why use `serverTimestamp()` here instead of `new_timestamp` from above?

JonathonYY replied:
It's likely irrelevant, but there's a slight delay introduced by `await updateDoc`, so I thought it best to have the listing use `serverTimestamp()`

Edwinhe03 reviewed (changes requested):
Great error validation and test coverage. Need minor changes and some clarification on rating behavior, then leaving some TODOs for a follow up PR.

JonathonYY committed:
Expand tests, input validation

Edwinhe03 reviewed (approved):
LGTM
